packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1970 @@
+   */
+  private validateCaveat(
+    caveat: CaveatConstraint,
+    origin: OriginString,
+    target: string,
+  ): void {
+    const specification = this.getCaveatSpecification(caveat.type);
+
+    if (!specification) {
+      throw new UnrecognizedCaveatTypeError(caveat.type, origin, target);
+    }
+
+    specification.validator?.(caveat, origin, target);
+  }
+
+  /**
+   * Initiates a permission request that requires user approval. This should
+   * always be used to grant additional permissions to a subject, unless user
+   * approval has been obtained through some other means.
+   *
+   * Permissions are validated at every step of the approval process, and this
+   * method will reject if validation fails.
+   *
+   * @see {@link ApprovalController} For the user approval logic.
+   * @see {@link PermissionController.acceptPermissionsRequest} For the method
+   * that _accepts_ the request and resolves the user approval promise.
+   * @see {@link PermissionController.rejectPermissionsRequest} For the method
+   * that _rejects_ the request and the user approval promise.
+   * @param origin - The origin of the grantee subject.
+   * @param requestedPermissions - The requested permissions.
+   * @param id - The id of the permissions request.
+   * @param preserveExistingPermissions - Whether to preserve the subject's
+   * existing permissions.
+   * @returns The granted permissions and request metadata.
+   */
+  async requestPermissions(
+    origin: OriginString,
+    requestedPermissions: RequestedPermissions,
+    id: string = nanoid(),
+    preserveExistingPermissions = true,
+  ): Promise<
+    [
+      Record<
+        string,
+        ExtractPermission<
+          ControllerPermissionSpecification,
+          ControllerCaveatSpecification
+        >
+      >,
+      { id: string; origin: OriginString },
+    ]
+  > {
+    this.validateRequestedPermissions(origin, requestedPermissions);
+    const metadata = {
+      id,
+      origin,
+    };
+
+    const permissionsRequest = {
+      metadata,
+      permissions: requestedPermissions,
+    };
+
+    const { permissions: approvedPermissions, ...requestData } =
+      await this.requestUserApproval(permissionsRequest);
+
+    return [
+      // Further validation is performed here
+      this.grantPermissions({
+        subject: { origin },
+        approvedPermissions,
+        preserveExistingPermissions,
+        requestData,
+      }),
+      metadata,
+    ];
+  }
+
+  /**
+   * Validates requested permissions. Throws if validation fails.
+   *
+   * This method ensures that the the requested permissions are a properly
+   * formatted {@link RequestedPermissions} object, that its keys correspond to
+   * the `parentCapability` fields of its values, and that the requested
+   * capability exists.
+   *
+   * Unrecognzied properties on requested permissions are ignored. Any requested
+   * caveats are validated in later stages of request processing.
```

Comment:
I see that they're validated during `constructCaveat`, which is called as part of `grantPermissions`. I was going to suggest that we explain that here, but.... I think we should be validating them here as well, before we ask the UI to render the requested caveats.

Otherwise the UI might be asked to display something it doesn't understand. The best we could do in that circumstance is throw an error, and better that we do that earlier on before bothering the user with a popup.

Reply:
It seems appropriate for us to validate both the requested permissions (from dapps, as they are untrusted), the approved permissions (from our UI, in case it gets compromised or has a bug), and the permissions being granted in `grantPermissions` (in case we call `grantPermissions` directly internally and make a mistake or are compromised).

Arguably the "approved permission" validation is redundant next to the "grant permission" validation, so I can see us skipping some validation steps there maybe. But validation of the requested permissions here would not be redundant.

Reply:
Validation updated here: 36efcf7

There is a lot of redundancy in our validation at the moment, mostly in `requestPermissions`, but we can revisit that later. Performance-wise, it should be completely fine.